drivers/acpi/acpica/osl.c

```diff
@@ -246,16 +262,44 @@ BOOLEAN AcpiOsWriteable(void *Memory, ACPI_SIZE Length) {
     return success;
 }
 
+static inline mapped_frame_t *find_mapped_frame(mfn_t mfn) {
+    mapped_frame_t *frame;
+
+    list_for_each_entry (frame, &mapped_frames, list) {
+        if (frame->mfn == mfn)
+            return frame;
+    }
+
+    return NULL;
+}
+
+static inline void new_mapped_frame(mfn_t mfn) {
+    mapped_frame_t *frame = kzalloc(sizeof(*frame));
+    frame->mfn = mfn;
+    frame->refcount = 1;
+    list_add(&frame->list, &mapped_frames);
+}
+
 void *AcpiOsMapMemory(ACPI_PHYSICAL_ADDRESS PhysicalAddress, ACPI_SIZE Length) {
     unsigned offset = PhysicalAddress & ~PAGE_MASK;
     unsigned num_pages = ((offset + Length) / PAGE_SIZE) + 1;
     mfn_t mfn = paddr_to_mfn(PhysicalAddress);
     void *va = NULL;
 
     for (unsigned i = 0; i < num_pages; i++, mfn++) {
-        void *_va = mmap_4k(mfn, L1_PROT);
-        if (!_va)
-            return NULL;
+        mapped_frame_t *frame = find_mapped_frame(mfn);
+        void *_va;
+
+        if (!frame) {
+            _va = mmap_4k(mfn, L1_PROT);
+            if (!_va)
+                return NULL;
+            new_mapped_frame(mfn);
+        }
+        else {
+            frame->refcount++;
```

Comment:
Don't need any lock?

Reply:
In theory yes. In practice no. The ACPICA runs only on the BSP. It may schedule functions on other CPUs, but we currently do not use it.

Reply:
Added locking.